grafana/grafana
# Loki: backend: use custom http code

the grafana Loki backend code currently uses a third-party package to do:
1. do the loki http request
2. parse the JSON response-bytes into a go data structure

unfortunately the third-party package is not able to do certain things we need (cancel the request and set custom http header). so this PR changes things so that:
1. we handle the loki http request in grafana code
2. this step is the same as it was before


how to test:
- first we test that it returns metric responses in a reasonable way
    - make sure your loki datasource works, go to explore mode
    - (this is a temporarily problem, will go away soon), switch to INSTANT-mode, then back to RANGE-mode
    - run a metric query like `count_over_time({level="info"}[10s])`
    - verify that the query is going through the backend-mode by observing the AJAX query in the browser-developer-tools
        - it should go to `/api/ds/query`, not to `/api/datasources/proxy/...`
    - verify that this returns a graph that looks ok
- we will test that when a query is running in the browser, and you press the [CANCEL] button, the http-request between go_server<>loki is cancelled too
    - make sure you have the feature-flag enabled, have this in your `conf/custom.ini`:
    ```
    [feature_toggles]
    enable = lokiBackendMode
    ```
    - we will run a "fake loki" server, a nodejs script that always returns an empty metric response after waiting 4 seconds, and it will show debug info for us:
    ```
    const http = require('http');

    const listener = (req, res) => {
      console.log('request received');
      const token = setTimeout(() => {
        res.writeHead(200);
        res.end(`{"status": "success","data": {"resultType": "matrix","result": []}}`);
        console.log('response sent');
      }, 4 * 1000);

      req.on('close', () => {console.log('request closed');});
      req.on('end', () => {console.log('request ended');});

      req.on('error', (e) => {
        console.log('error', e);
        clearTimeout(token);
      });
    };

    const server = http.createServer(listener);
    server.listen(3100, '127.0.0.1');
    ```
    - turn off your "localhost loki" if you are running any
    - run the script  with `node server.js`
    - make sure your loki datasource points to `localhost:3100` (or adjust the script)
    - go to explore-mode
    - (this is a temporarily problem, will go away soon), switch to INSTANT-mode, then back to RANGE-mode
    - enter a metric loki query, like `count_over_time({level="info"}[10s)`
    - run the query, it should finish in 4 seconds. observe the output from the nodejs script, it should be roughly:
    ```
    request received
    response sent
    request ended
    request closed
    ```
    - (sometimes we run additional loki queries to load metadata, so you might see duplicated log lines)
    - verify that the query is going through the backend-mode by observing the AJAX query in the browser-developer-tools
        - it should go to `/api/ds/query`, not to `/api/datasources/proxy/...`
    - now do the same query, but while it is running, press [CANCEL]
    - you should see roughly this in the nodejs script's output:
    ````
    request received
    error Error: aborted
        .
        .
        .
        {code: 'ECONNRESET'}
    request closed
    ````
